api/solarperformanceinsight_api/compute.py

```diff
@@ -444,3 +507,293 @@ def compare_expected_and_actual(job: models.StoredJob, si: storage.StorageInterf
+        # converting irradiance to POA. In the future, could only run these conversions
+        # instead of full modelchain calculation.
+        # since it sets class properties, copy for the actuals calculations
+        chain_actual = deepcopy(chain)
+        inv = job.definition.system_definition.inverters[i]
+        pac0 = inv.inverter_parameters._pac0
+        num_arrays = len(chain.system.arrays)
+        gammas = [
+            arr._gamma for arr in job.definition.system_definition.inverters[i].arrays
+        ]
+        if any([g is None for g in gammas]):
+            raise TypeError(
+                "Currently unable to compare predicted and actual performance for "
+                "PVsyst specified arrays."
+            )
+        if data_available == models.PredictedDataEnum.weather_only:  # 2A-4
+            db_results, _ = process_single_modelchain(
+                chain, ref_weather, ref_model_method, tshift, i
+            )
+            results_list += db_results
+            ref_pdc = adjust(_get_mc_dc(chain.results, num_arrays))  # type: ignore
+        else:
+            # run chain on ref weather
+            getattr(chain, ref_model_method)(
+                [d.shift(freq=tshift) for d in ref_weather]  # type: ignore
+            )
+
+        # run chain on actual weather
+        getattr(chain_actual, actual_model_method)(
+            [d.shift(freq=tshift) for d in actual_weather]  # type: ignore
+        )
+        # use pvlib.modelchain._irrad_for_celltemp that returns POA global if available
+        # otherwise uses effective irradiance
+        poa_ref = _irrad_for_celltemp(
+            chain.results.total_irrad, chain.results.effective_irradiance
+        )
+        poa_actual = _irrad_for_celltemp(
+            chain_actual.results.total_irrad, chain_actual.results.effective_irradiance
+        )
+        # use cell temperature from the pvlib modelchain
+        # If air temp + wind speed were supplied, they are converted
+        # to cell temperature via the array's temperature model.  If
+        # module_temperature was supplied and the temperature model is
+        # sapm, it is converted to cell_temperature, otherwise module
+        # temperature is used in place of cell_temperature
+        t_ref = _get_temp(
+            ref_weather, chain.results.cell_temperature, inv.arrays, tshift
+        )
+        t_actual = _get_temp(
+            actual_weather, chain_actual.results.cell_temperature, inv.arrays, tshift
+        )
+
+        # mean of array POArat * TempFactor for this inverter
+        # could make more sense to use weighted mean with weights set
+        # by array power percentage
+        # another alternative, calculate average POArat and TempFactor separately
+        poa_rat = [_zero_div(pa, pr) for pa, pr in zip(poa_actual, poa_ref)]
+        tempfactor = list(map(_temp_factor, gammas, t_ref, t_actual))
```

Comment:
Is there a reason you're using a list comprehenion and then a map? The style difference threw me off.

Reply:
not particularly. `_zero_div` came later, before it was just `[ pa / pr for ...]` so the comprehension made more sense